File: daily_problems/2025/01/0127/personal_submission/cf374c_hang.cpp
#include <bits/stdc++.h>
using namespace std;

int n,m;
vector<string> g;
vector<vector<int>> dp, vis;
string nxt = "DIMA";

int dfs(int x,int y){
    if(dp[x][y] != -1) return dp[x][y];
    if(vis[x][y] == 1){
        cout<<"Poor Inna!";
        exit(0);
    }
    vis[x][y] = 1;
    int res=0;
    vector<pair<int,int>> dirs={{1,0},{-1,0},{0,1},{0,-1}};
    char need = nxt[(nxt.find(g[x][y]) + 1) % 4];
    for(auto &d:dirs){
        int nx=x+d.first, ny=y+d.second;
        if(nx>=0 && nx<n && ny>=0 && ny<m && g[nx][ny]==need){
            res = max(res, dfs(nx,ny));
        }
    }
    if(g[x][y]=='A') res++;
    dp[x][y]=res; 
    vis[x][y] = 0;
    return res;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin>>n>>m;
    g.resize(n);
    for(int i=0;i<n;i++){
        cin>>g[i];
    }
    dp.assign(n, vector<int>(m,-1));
    vis.assign(n, vector<int>(m,0));
    int ans=0;
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(g[i][j]=='D'){
                ans = max(ans, dfs(i,j));
            }
        }
    }
    if(ans==0) cout<<"Poor Dima!";
    else cout<<ans;
    return 0;
}

File: daily_problems/2025/01/0127/personal_submission/cf374c_hang.py
# 过不去，需要cpp实现
def main():
    n,m = map(int,input().split())
    g = []
    for _ in range(n):
        g.append(input())
    vis = [[0] * m for _ in range(n)]
    dp = [[-1] * m for _ in range(n)]
    nxt = {'D': 'I', 'I': 'M', 'M': 'A', 'A': 'D'}
    dirs = [(1,0),(-1,0),(0,1),(0,-1)]
    ans = 0
    cur = 0

    def dfs(x,y):
        if dp[x][y] != -1:
            return dp[x][y]
        if vis[x][y]:
            exit(print('Poor Inna!'))
        vis[x][y] = 1
        res = 0
        
        for dx,dy in dirs:
            nx,ny = x+dx,y+dy
            if 0 <= nx < n and 0 <= ny < m and g[nx][ny] == nxt[g[x][y]]:
                res = max(dfs(nx,ny), res)
        if g[x][y] == 'A':
            res += 1
        vis[x][y] = 0
        dp[x][y] = res
        return res
    for i in range(n):
        for j in range(m):
            if g[i][j] == 'D':
                cur = dfs(i,j)
                ans = max(ans,cur)
    if ans == 0:
        print('Poor Dima!')
    else:
        print(ans)
                
main()